New assert API doesn't support ClientFunction as an argument in the expect function

I can pass the Selector object to the expect function, but when I pass ClientFunction I get the following error in the output:
`AssertionError: expected [Function: __$$clientFunction$$] to deeply equal 2`

--- a/src/api/assertion.js
+++ b/src/api/assertion.js
@@ -3,7 +3,7 @@ import { assign } from 'lodash';
 import delay from '../utils/delay';
 import { ExternalAssertionLibraryError } from '../errors/test-run';
 import { assertNonNegativeNumber } from '../errors/runtime/type-assertions';
-import SelectorResultPromise from '../client-functions/selector-builder/result-promise';
+import ClientFunctionResultPromise from '../client-functions/result-promise';
 
 const ASSERTION_DELAY = 200;
 
@@ -43,7 +43,7 @@ export default class Assertion {
     }
 
     _wrapExecutor (callsite, executor) {
-        if (this.actual instanceof SelectorResultPromise)
+        if (this.actual instanceof ClientFunctionResultPromise)
             executor = this._wrapSelectorResultAssertionExecutor(executor);
 
         return async () => {

--- a/src/client-functions/client-function-builder.js
+++ b/src/client-functions/client-function-builder.js
@@ -10,6 +10,7 @@ import { assertObject, assertNonNullObject } from '../errors/runtime/type-assert
 import MESSAGE from '../errors/runtime/message';
 import getCallsite from '../errors/get-callsite';
 import ensureDeprecatedOptions from './selector-builder/ensure-deprecated-options';
+import ClientFunctionResultPromise from './result-promise';
 
 const DEFAULT_EXECUTION_CALLSITE_NAME = '__$$clientFunction$$';
 
@@ -116,14 +117,29 @@ export default class ClientFunctionBuilder {
     }
 
     _executeCommand (args, testRun, callsite) {
-        if (!testRun)
-            throw new ClientFunctionAPIError(this.callsiteNames.execution, this.callsiteNames.instantiation, MESSAGE.clientFunctionCantResolveTestRun);
-
+        // NOTE: should be kept outside of lazy promise to preserve
+        // correct callsite in case of replicator error.
         var command = this.getCommand(args);
 
-        return testRun
-            .executeCommand(command, callsite)
-            .then(result => this.replicator.decode(result));
+        return ClientFunctionResultPromise.fromFn(async () => {
+            if (!testRun) {
+                var err = new ClientFunctionAPIError(this.callsiteNames.execution, this.callsiteNames.instantiation, MESSAGE.clientFunctionCantResolveTestRun);
+
+                // NOTE: force callsite here, because more likely it will
+                // be impossible to resolve it by method name from a lazy promise.
+                err.callsite = callsite;
+
+                throw err;
+            }
+
+            var result = await testRun.executeCommand(command, callsite);
+
+            return this._processResult(result, args);
+        });
+    }
+
+    _processResult (result) {
+        return this.replicator.decode(result);
     }
 
     _validateOptions (options) {

--- a/src/client-functions/result-promise.js
+++ b/src/client-functions/result-promise.js
@@ -1,9 +1,9 @@
 import Promise from 'pinkie';
 import { noop } from 'lodash';
-import testRunTracker from '../test-run-tracker';
+import testRunTracker from './test-run-tracker';
 
 
-export default class SelectorResultPromise extends Promise {
+export default class ClientFunctionResultPromise extends Promise {
     constructor (executorFn) {
         super(noop);
 
@@ -40,7 +40,7 @@ export default class SelectorResultPromise extends Promise {
         if (testRunId)
             asyncExecutorFn = testRunTracker.addTrackingMarkerToFunction(testRunId, asyncExecutorFn);
 
-        return new SelectorResultPromise(resolve => resolve(asyncExecutorFn()));
+        return new ClientFunctionResultPromise(resolve => resolve(asyncExecutorFn()));
     }
 }
 

--- a/src/client-functions/selector-builder/add-api.js
+++ b/src/client-functions/selector-builder/add-api.js
@@ -4,7 +4,7 @@ import { ELEMENT_SNAPSHOT_PROPERTIES, NODE_SNAPSHOT_PROPERTIES } from './snapsho
 import { CantObtainInfoForElementSpecifiedBySelectorError } from '../../errors/test-run';
 import getCallsite from '../../errors/get-callsite';
 import ClientFunctionBuilder from '../client-function-builder';
-import SelectorResultPromise from './result-promise';
+import ClientFunctionResultPromise from '../result-promise';
 import {
     assertStringOrRegExp,
     assertNumber,
@@ -94,7 +94,7 @@ function addSnapshotPropertyShorthands (obj, getSelector) {
             get: () => {
                 var callsite = getCallsite('get');
 
-                return SelectorResultPromise.fromFn(async () => {
+                return ClientFunctionResultPromise.fromFn(async () => {
                     var snapshot = await getSnapshot(getSelector, callsite);
 
                     return snapshot[prop];
@@ -106,7 +106,7 @@ function addSnapshotPropertyShorthands (obj, getSelector) {
     obj.getStyleProperty = prop => {
         var callsite = getCallsite('getStyleProperty');
 
-        return SelectorResultPromise.fromFn(async () => {
+        return ClientFunctionResultPromise.fromFn(async () => {
             var snapshot = await getSnapshot(getSelector, callsite);
 
             return snapshot.style ? snapshot.style[prop] : void 0;
@@ -116,7 +116,7 @@ function addSnapshotPropertyShorthands (obj, getSelector) {
     obj.getAttribute = attrName => {
         var callsite = getCallsite('getAttribute');
 
-        return SelectorResultPromise.fromFn(async () => {
+        return ClientFunctionResultPromise.fromFn(async () => {
             var snapshot = await getSnapshot(getSelector, callsite);
 
             return snapshot.attributes ? snapshot.attributes[attrName] : void 0;
@@ -126,7 +126,7 @@ function addSnapshotPropertyShorthands (obj, getSelector) {
     obj.getBoundingClientRectProperty = prop => {
         var callsite = getCallsite('getBoundingClientRectProperty');
 
-        return SelectorResultPromise.fromFn(async () => {
+        return ClientFunctionResultPromise.fromFn(async () => {
             var snapshot = await getSnapshot(getSelector, callsite);
 
             return snapshot.boundingClientRect ? snapshot.boundingClientRect[prop] : void 0;
@@ -136,7 +136,7 @@ function addSnapshotPropertyShorthands (obj, getSelector) {
     obj.hasClass = name => {
         var callsite = getCallsite('hasClass');
 
-        return SelectorResultPromise.fromFn(async () => {
+        return ClientFunctionResultPromise.fromFn(async () => {
             var snapshot = await getSnapshot(getSelector, callsite);
 
             return snapshot.classNames ? snapshot.classNames.indexOf(name) > -1 : false;
@@ -166,7 +166,7 @@ function addCounterProperties (obj, getSelector, SelectorBuilder) {
         get: () => {
             var counter = createCounter(getSelector, SelectorBuilder);
 
-            return SelectorResultPromise.fromFn(() => counter());
+            return ClientFunctionResultPromise.fromFn(() => counter());
         }
     });
 
@@ -174,7 +174,7 @@ function addCounterProperties (obj, getSelector, SelectorBuilder) {
         get: () => {
             var counter = createCounter(getSelector, SelectorBuilder);
 
-            return SelectorResultPromise.fromFn(async () => await counter() > 0);
+            return ClientFunctionResultPromise.fromFn(async () => await counter() > 0);
         }
     });
 }

--- a/src/client-functions/selector-builder/index.js
+++ b/src/client-functions/selector-builder/index.js
@@ -6,13 +6,17 @@ import { ClientFunctionAPIError } from '../../errors/runtime';
 import functionBuilderSymbol from '../builder-symbol';
 import MESSAGE from '../../errors/runtime/message';
 import getCallsite from '../../errors/get-callsite';
-import { assertNumber, assertNonNegativeNumber, assertBoolean, assertStringOrRegExp } from '../../errors/runtime/type-assertions';
+import {
+    assertNumber,
+    assertNonNegativeNumber,
+    assertBoolean,
+    assertStringOrRegExp
+} from '../../errors/runtime/type-assertions';
 import deprecate from '../../warnings/deprecate';
 import { ExecuteSelectorCommand } from '../../test-run/commands/observation';
 import defineLazyProperty from '../../utils/define-lazy-property';
 import addAPI from './add-api';
 import createSnapshotMethods from './create-snapshot-methods';
-import SelectorResultPromise from './result-promise';
 
 export default class SelectorBuilder extends ClientFunctionBuilder {
     constructor (fn, options, callsiteNames) {
@@ -95,18 +99,14 @@ export default class SelectorBuilder extends ClientFunctionBuilder {
     }
 
     _executeCommand (args, testRun, callsite) {
-        var lazyPromise = SelectorResultPromise.fromFn(async () => {
-            var result = await super._executeCommand(args, testRun, callsite);
+        var resultPromise = super._executeCommand(args, testRun, callsite);
 
-            return result && !this.options.counterMode ? this._decorateFunctionResult(result, args) : result;
-        });
-
-        this._addBoundArgsSelectorGetter(lazyPromise, args);
+        this._addBoundArgsSelectorGetter(resultPromise, args);
 
         // OPTIMIZATION: use buffer function as selector not to trigger lazy property ahead of time
-        addAPI(lazyPromise, () => lazyPromise.selector, SelectorBuilder);
+        addAPI(resultPromise, () => resultPromise.selector, SelectorBuilder);
 
-        return lazyPromise;
+        return resultPromise;
     }
 
     getFunctionDependencies () {
@@ -177,39 +177,43 @@ export default class SelectorBuilder extends ClientFunctionBuilder {
         addAPI(selectorFn, () => selectorFn, SelectorBuilder);
     }
 
-    _decorateFunctionResult (nodeSnapshot, selectorArgs) {
-        this._addBoundArgsSelectorGetter(nodeSnapshot, selectorArgs);
+    _processResult (result, selectorArgs) {
+        var snapshot = super._processResult(result, selectorArgs);
 
-        SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(nodeSnapshot, 'getParentNode', node => {
-            return node ? node.parentNode : node;
-        });
+        if (snapshot && !this.options.counterMode) {
+            this._addBoundArgsSelectorGetter(snapshot, selectorArgs);
 
-        SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(nodeSnapshot, 'getChildNode', (node, idx) => {
-            return node ? node.childNodes[idx] : node;
-        });
+            SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(snapshot, 'getParentNode', node => {
+                return node ? node.parentNode : node;
+            });
 
-        SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(nodeSnapshot, 'getChildElement', (node, idx) => {
-            if (node.children)
-                return node.children[idx];
+            SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(snapshot, 'getChildNode', (node, idx) => {
+                return node ? node.childNodes[idx] : node;
+            });
 
-            // NOTE: IE doesn't have `children` for non-element nodes =/
-            var childNodeCount = node.childNodes.length;
-            var currentElIdx   = 0;
+            SelectorBuilder._defineNodeSnapshotDerivativeSelectorProperty(snapshot, 'getChildElement', (node, idx) => {
+                if (node.children)
+                    return node.children[idx];
 
-            for (var i = 0; i < childNodeCount; i++) {
-                if (node.childNodes[i].nodeType === 1) {
-                    if (currentElIdx === idx)
-                        return node.childNodes[i];
+                // NOTE: IE doesn't have `children` for non-element nodes =/
+                var childNodeCount = node.childNodes.length;
+                var currentElIdx   = 0;
 
-                    currentElIdx++;
+                for (var i = 0; i < childNodeCount; i++) {
+                    if (node.childNodes[i].nodeType === 1) {
+                        if (currentElIdx === idx)
+                            return node.childNodes[i];
+
+                        currentElIdx++;
+                    }
                 }
-            }
 
-            return null;
-        });
+                return null;
+            });
 
-        createSnapshotMethods(nodeSnapshot);
+            createSnapshotMethods(snapshot);
+        }
 
-        return nodeSnapshot;
+        return snapshot;
     }
 }

--- a/test/functional/fixtures/api/es-next/assertions/pages/index.html
+++ b/test/functional/fixtures/api/es-next/assertions/pages/index.html
@@ -45,6 +45,18 @@
             div.textContent = '42';
         }, 180);
     });
+
+    var value = 4;
+
+    Object.defineProperty(window, 'someVar', {
+        get: function () {
+            window.setTimeout(function () {
+                value = 2;
+            }, 400);
+
+            return value;
+        }
+    })
 </script>
 
 </body>

--- a/test/functional/fixtures/api/es-next/assertions/test.js
+++ b/test/functional/fixtures/api/es-next/assertions/test.js
@@ -222,4 +222,9 @@ describe('[API] Assertions', function () {
             assertionTimeout: 0
         });
     });
+
+    it('Should retry assertion for ClientFunction results', function () {
+        return runTests('./testcafe-fixtures/assertions-test.js', 'ClientFunction result assertion', { only: 'chrome' });
+    });
+
 });

--- a/test/functional/fixtures/api/es-next/assertions/testcafe-fixtures/assertions-test.js
+++ b/test/functional/fixtures/api/es-next/assertions/testcafe-fixtures/assertions-test.js
@@ -1,4 +1,4 @@
-import { Selector } from 'testcafe';
+import { Selector, ClientFunction } from 'testcafe';
 fixture `Assertions`
     .page `http://localhost:3000/fixtures/api/es-next/assertions/pages/index.html`;
 
@@ -140,4 +140,10 @@ test('.notMatch() assertion', async t => {
     await t
         .expect('yo').notMatch(/[x,z]o/)
         .expect('42 hey').notMatch(/\d+ hey/);
+});
+
+test('ClientFunction result assertion', async t => {
+    const getSomeVar = ClientFunction(() => window.someVar);
+
+    await t.expect(getSomeVar()).eql(2);
 });

--- a/test/functional/fixtures/api/es-next/client-function/test.js
+++ b/test/functional/fixtures/api/es-next/client-function/test.js
@@ -108,7 +108,7 @@ describe('[API] ClientFunction', function () {
                     'ClientFunction cannot implicitly resolve the test run in context of which it should be executed.'
                 )).eql(0);
 
-                expect(errs[0]).contains(' > 43 |                fn();');
+                expect(errs[0]).contains(' > 43 |                await fn();');
             });
         });
 

--- a/test/functional/fixtures/api/es-next/client-function/testcafe-fixtures/client-fn-test.js
+++ b/test/functional/fixtures/api/es-next/client-function/testcafe-fixtures/client-fn-test.js
@@ -38,9 +38,9 @@ test('ClientFunction fn test run is unresolvable', async () => {
     var fn = ClientFunction(() => 123);
 
     return new Promise((resolve, reject) => {
-        fs.readFile('not/exists', () => {
+        fs.readFile('not/exists', async () => {
             try {
-                fn();
+                await fn();
                 resolve();
             }
             catch (err) {

--- a/test/server/compiler-test.js
+++ b/test/server/compiler-test.js
@@ -680,37 +680,6 @@ describe('Compiler', function () {
                 });
         });
 
-
-        it('Should raise an error if ClientFunction not able to resolve test run', function () {
-            var testfile = resolve('test/server/data/test-suites/client-fn-cant-resolve-test-run/testfile.js');
-
-            return compile(testfile)
-                .then(function () {
-                    throw new Error('Promise rejection expected');
-                })
-                .catch(function (err) {
-                    assertAPIError(err, {
-                        stackTop: testfile,
-
-                        message: 'Cannot prepare tests due to an error.\n\n' +
-                                 'ClientFunction cannot implicitly resolve the test run in context of which it ' +
-                                 'should be executed. If you need to call ClientFunction from the Node.js API ' +
-                                 "callback, pass the test controller manually via ClientFunction's `.with({ boundTestRun: t })` " +
-                                 'method first. Note that you cannot execute ClientFunction outside the test code.',
-
-                        callsite: "   1 |import { ClientFunction } from 'testcafe';\n" +
-                                  '   2 |\n' +
-                                  '   3 |fixture `Test`;\n' +
-                                  '   4 |\n' +
-                                  ' > 5 |ClientFunction(() => 123)();\n' +
-                                  '   6 |\n' +
-                                  "   7 |test('yo', () => {\n" +
-                                  '   8 |});\n' +
-                                  '   9 |'
-                    });
-                });
-        });
-
         it('Should raise an error if ClientFunction uses async function', function () {
             var testfile = resolve('test/server/data/test-suites/async-function-in-client-fn/testfile.js');
 

--- a/test/server/data/test-suites/client-fn-cant-resolve-test-run/testfile.js
+++ b/test/server/data/test-suites/client-fn-cant-resolve-test-run/testfile.js
@@ -1,8 +1,0 @@
-import { ClientFunction } from 'testcafe';
-
-fixture `Test`;
-
-ClientFunction(() => 123)();
-
-test('yo', () => {
-});

